### DISCUSSION TITLE
Looking for advice on INotifyDataErrorInfo implementation

### DISCUSSION THREAD

**Discussion by kcabral817** [MarimerLLC/csla#1675] [Questions] [answered]
I am working on getting the brokenRules to notify the WPF xaml code to display the red box around the control when the value is invalid.  I read that I should implement INotifyDataErrorInfo and grab the collection of errors returned.  I have a complex object that is displayed flat in the UI and I'm not getting the controls highlighted when the control is skipped and is required even though I call ((ICheckRules)Model).CheckRules();.   If the control has an invalid value entered, it works well (without implementing INotifyDataErrorInfo in the class.)  Here are my questions, if you'd be so kind to weigh in:

1) Should I implement INotifyDataErrorInfo as appropriate?
2) Is there a way to consolidate all the errors from the children (I read a archived piece from Jason Brock but it's 10 years old).
3) Should I use ((ICheckRules)Model).CheckRules(); or just try and save and get the errors back from the BO (I read the encapsulation warning in the CSLA book)
4) Is there an example someone could point me towards?

If I implement INotifyDataErrorInfo, it's a couple of days worth of work and I just trying to make sure my efforts are fruitful.  

Thank you for taking the time to read this.

**Comment by rockfordlhotka** [comment]
CSLA already implements the correct interface(s) for data binding, along with all the somewhat complex behaviors required to make them actually work correctly.

Why do you need to implement them yourself?

**Reply by kcabral817** [reply]
When the brokenRules collection is returned for the an object, the red borders do not show up around the controls that have no data.  They only show when the user enters an invalid value, not when they miss one.  I'm trying to catch that situation where the user fails to enter any value and needs a visual queue to identify the offending control.  The rules are firing and I'm getting the rule list back, but the adorners are not getting set for controls missed.

**Reply by rockfordlhotka** [reply]
Using standard WPF binding, or using `Csla.Xaml.PropertyInfo`?

**Reply by kcabral817** [reply]
Yes.  I have the following in the XAML file:

```xml
 <Label Content="Manufactured Date:" Grid.Column="0" Grid.Row="10" />
                <igEditors:XamDateTimeEditor x:Name="ManufacturedDtXamDateTimeEditor" 
                                             Mask="{}{date}"
                                             Grid.Column="1" Grid.Row="10" 
                                             Value="{Binding Model.ManufacturedDt, Mode=TwoWay, ValidatesOnDataErrors=True, ValidatesOnNotifyDataErrors =True, NotifyOnValidationError=true, NotifyOnSourceUpdated=True, NotifyOnTargetUpdated=True, ValidatesOnExceptions =true}" />
                <csla:PropertyStatus Property="{Binding Path=Model.ManufacturedDt}" Grid.Column="1" Grid.Row="10"  HorizontalAlignment="Left" />
```

The validation works if an entry is made, but if it is skipped, the broken rule does not get highlighted nor displayed although the broken rule is in the collection.  Thoughts?

**Reply by kcabral817** [reply]
I've been working on this, and thought I might want to implement:

```c#
 public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
        private void RaiseErrorsChanged(string propertyName)
        {
            if (ErrorsChanged != null)
                ErrorsChanged(this, new DataErrorsChangedEventArgs(propertyName));
        }
```

I'm thinking that if I iterate through the collection and raise the event, the binding would pick it up.   However,  I can't find INotifyDataErrorInfo when I search through CSLA.

**Comment by kcabral817** [comment]
I'm really lost on this.  I have the brokenRules collection and used  BusinessRules.GetAllBrokenRules(Model); to consolidate them.  I have 16 errors, but none of them display next to the control such as when the user enters an invalid value--that works wonderfully.  The missed field, not so much.

**Comment by rockfordlhotka** [comment]
I still don't fully understand the issue you are trying to solve.

Are you saying:

1. You have a class with a `Required` property
1. You create a new instance of the type and bind it to the UI
1. `PropertyStatus` doesn't show the broken rule
1. After you edit the property, then `PropertyStatus` shows broken rules

???

**Reply by kcabral817** [reply]
Thank you for the questions, they lead me into a direction of the business rule invocation.  I just need some advice on how to do,  what I think is a gateway rule, for a peer class.  I have a class that needs to call the validations of a "peer" class.   This is where my problem actually lies.  I called the peer's class using (ICheckRules)Model.ElectricMeter).CheckRules(); from the UI (who's model = EquipmentItem) which doesn't work.  I would like to check the Electric meter's rules through the gateway but I'm not sure how to pass the electric meter object so it invokes the rules.

**Comment by kcabral817** [comment, marked as answer]
In my factory class, I added a CheckRules for the child and all is well.  Presents dropped from the sky tonight!
